WIP: Documentation for Vault Swaps (#471)
chainflip-io/chainflip-docs: pages/swapping/integrations/advanced/vault-swaps/vault-swaps.mdx

2. Once the transaction is obtained, this must be signed and submitted to the correct smart contract address.

3. Once the smart contract receives the transaction, funds are automatically transferred and the swap is processed.

## Constructing the Vault Swap Transaction

Each of our supported chains encodes data differently. To make this feature easy to use, Chainflip provides the interface to build the transaction on the integrator's behalf. This way the transaction only needs to be signed and is then ready to be sent out to the perspective Vault contract.

The function that builds the encoded the transaction data is the [request_swap_parameter_encoding](../../running-a-broker/broker-api/#broker_request_swap_parameter_encoding) RPC call in the Broker Api.

<Callout type="info">
The function `request_swap_parameter_encoding()` looks very similar to `request_swap_deposit_address()`, but they are very different functions. 

`request_swap_deposit_address()` will submit an extrinsic that opens a deposit channels, and will return the deposit address.

> `request_swap_parameter_encoding()` will **not** make any on-chain changes, but will instead builds a transaction that will be returned.

Reviewer: ```suggestion
`request_swap_parameter_encoding()` will **not** make any on-chain changes, but will instead build a transaction that will be returned.
```

builds -> build

Traditionally swaps initiated by brokers typically involves requesting an deposit address, then deposit fund into that deposit address to initiate the swap. This page introduces a new way to initiating swaps on the Chainflip network, where a direct call to the Chainflip Vault is used.

Vault swaps can be started without the use of [Deposit Channel/Addresses](../javascript-sdk/swap-assets/request-deposit-address/v2). Vault swaps support the same features as the swapps via deposit addresses.

<Callout type="info">
For more information, learn [How Swapping Works](../../how-swapping-works.mdx).
</Callout>

## Workflows

### Swap via Deposit Channels

1. Request a deposit channel via the `pallet_cf_swapping::request_swap_deposit_address_with_affiliates()` extrinsic. This can be done directly or via the `request_swap_deposit_address` RPC call from our Broker Api.

2. Deposit fund into the given deposit address.

3. Swaps are processed after deposits are witnessed by Chainflip Engine

### Vault swaps

1. Construct and submit a transaction on an external chain. This will be a smart contract call for EVM chains, a program call for Solana, or a particularly encoded transaction for Bitcoin.

> 2. This can be done by manually encoding the transaction (e.g. making the smart contract call) or by using the `request_swap_parameter_encoding` RPC call from our Broker Api. It's encouraged to use the broker API to ensure that the payloads are correctly encoded according to the specifications. Wrongly encoded payloads may result in lost of funds.

Reviewer: ```suggestion
2. This can be done by manually encoding the transaction (e.g. making the smart contract call) or by using the `request_swap_parameter_encoding` RPC call from our Broker Api. It's encouraged to use the broker API to ensure that the payloads are correctly encoded according to the specifications. Wrongly encoded payloads may result in loss of funds.
```

3. If using the Broker API, an encoded payload is obtained and must be signed and sent to the appropriate blockchain.

4. The transaction will transfer the appropriate funds to the Chainflip protocol as part of the smart contract call. Chainflip will then process the swap.

## Constructing a Vault Swap Transaction

> Each of our supported chains encodes data differently. To make this feature easy to use, Chainflip provides the interface to build the transaction on the integrator's behalf. This way the transaction only needs to be signed and is then ready to be sent out to the perspective Vault contract.

Reviewer: ```suggestion
Each of our supported chains encodes data differently. To make this feature easy to use, Chainflip provides the interface to build the transaction on the integrator's behalf. This way the transaction only needs to be signed and is then ready to be sent out to the respective Vault contract.
```

> The function that builds the encoded the transaction data is the [request_swap_parameter_encoding](../../running-a-broker/broker-api/#broker_request_swap_parameter_encoding) RPC call in the Broker Api.

Reviewer: ```suggestion
The function that builds the encoded the transaction data is the [request_swap_parameter_encoding](../../running-a-broker/broker-api/#broker_request_swap_parameter_encoding) RPC call in the Broker API.
```

        "is_writable": true
      },
      {
        "pubkey": "2tmtGLQcBd11BMiE9B1tAkQXwmPNgR79Meki2Eme4Ec9",
        "is_signer": false,
        "is_writable": true
      },
      {
        "pubkey": "11111111111111111111111111111111",
        "is_signer": false,
        "is_writable": false
      }
    ],
    "data": "0xa3265ce2f3698dc400ca9a3b000000000100000014000000c64722ad9613851b10e26ff8118a7696a0f956f201000000006b000000000a000000049e0d6a70e12d54edf90971cc977fa26a1d3bb4b0b26e72470171c36b0006b01f000000000000000000000000000000000000000000000000000000000000000000009059e6d854b769a505d01148af212bf8cb7f8469a7153edce8dcaedd9d299125000000"
  },
  "id": 1
}
```

<Callout type="info">
The transaction returned as result differs depending on the `source_asset`. The `chain` field will indicate the exact chain the transaction is built for. For the example above, since the source asset was Sol, the transaction built was for the Solana chain.
</Callout>

### Example with CCM

To request a vault Swap with [cross-chain messaging](../cross-chain-messaging/cross-chain-messaging.mdx) the `channel_metadata` field parameter must be filled. You can see an example below:

<Callout type="info">

> Currently only Solana and EVM chains supports CCM. This means **CCM is only allows with vault swaps with `output asset` being on the Solana or EVM chain (such as Sol or ArbEth)**. The mechanism differs between Solana and EVM chains, please refer to [CCM support](../cross-chain-messaging/cross-chain-messaging.mdx) for more details.

Reviewer: ```suggestion
Currently only Solana and EVM chains supports CCM. This means **CCM is only allowed with vault swaps with `output asset` being on the Solana or EVM chain (such as Sol or ArbEth)**. The mechanism differs between Solana and EVM chains, please refer to [CCM support](../cross-chain-messaging/cross-chain-messaging.mdx) for more details.
```

</Callout>

```sh
curl -H "Content-Type: application/json" -d '{
    "id": 1,
    "jsonrpc": "2.0",
    "method": "broker_request_swap_parameter_encoding",
    "params": [
		{ "chain": "Solana", "asset": "SOL" },
        { "chain": "Ethereum", "asset": "ETH" },
        "0xc64722AD9613851b10E26fF8118A7696A0f956f2",
        0,
        {
            "chain": "Solana",
            "from": "BdyHK5DckpAFGcbZveGLPjjMEaADGfNeqcRXKoyd33kA",
            "event_data_account": "9acHwMGmeoMr5o8Cw1V2U4HjMQwhced3eQP31yYEhYDU",
            "input_amount": 1000000000,
            "refund_parameters": {
                "retry_duration": 10,
                "refund_address": "BdyHK5DckpAFGcbZveGLPjjMEaADGfNeqcRXKoyd33kA",
                "min_price": "0x0"
            }
        },
        {
          "message": "0x0011223344556677",
          "gas_budget": 1000,
          "ccm_additional_data": "0x"
        }
    ]
}' http://localhost:10997
```

> ### Construct, sign and and broadcast a transaction

Reviewer: ```suggestion
### Construct, sign and broadcast a transaction
```

> With the payload encoded, the next step is construct a transaction from that payload and sent to the correct chain's Vault contract address. This step differs for each chain/asset. Please refer to the page dedicated for each chain for more details and other information.

Reviewer: ```suggestion
With the payload encoded, the next step is to construct a transaction from that payload and send it to the correct chain's Vault contract address. This step differs for each chain/asset. Please refer to the page dedicated for each chain for more details and other information.
```

</Callout>

[For Bitcoin network](./bitcoin.mdx)

[For Ethereum and Arbitrum](./evm.mdx)

[For Solana](./solana.mdx)

### Payload expiry

Payloads and therefore transactions for Vault swaps are valid for long periods of time and in some cases indefinitely. However, it is strongly recommended to only sign and broadcast freshly encode payloads and to never rely on previous unused payloads.

---

### References

#### Input parameters

| Parameter | Description | Data Type |
|-----------|-------------|-----------|
| source_asset | Input asset for the swap | `Asset` |
| destination_asset | Output asset for the swap | `Asset` |
| destination_address | Destination address the swap is paid out to | `AddressString` |
| broker_commission | Broker commission in Basis point | `u16` |
| extra_parameters | Chain specific extra parameters. For details see each chain's dedicated page | `VaultSwapExtraParametersRpc` |
| channel_metadata | Optional: Information for CCM | `Option<CcmChannelMetadata>` |
| boost_fee | Optional: Boost fee for the swap ingress  | `Option<u16>` |
| affiliate_fees | Optional: A list of affiliates and their fees | `Option<Affiliates<AccountId32>>` |

> | dca_parameters | Optional: Parameters for Dca swaps | `Option<DcaParameters>` |

Reviewer: ```suggestion
| dca_parameters | Optional: Parameters for DCA swaps | `Option<DcaParameters>` |
```

#### Type References 

The following are references for the Types used in Vault Swap calls. This can be used for references when building your RPC call.

```rust
/// For CCM message
pub struct CcmChannelMetadata {
	/// A Vec of bytes with maximum length of 15_000. `BoundedVec<u8, ConstU32<15_000>>`
	pub message: CcmMessage,
	/// User funds designated to be used for gas.
	pub gas_budget: GasAmount,
	/// Additional parameters for the cross chain message. Only used for Solana CCMs 
	/// A Vec of bytes with maximum length of 1_000. `BoundedVec<u8, ConstU32<1_000>>`
	pub ccm_additional_data: CcmAdditionalData,
}

/// Bounded vec of affiliates and fees in Basis Points. Max size of 5
pub type Affiliates<Id> = BoundedVec<Beneficiary<Id>, ConstU32<MAX_AFFILIATES>>;

pub struct Beneficiary<Id> {
	pub account: Id,
	pub bps: u16,
}

/// Parameters for DCA
pub struct DcaParameters {
	/// The number of individual swaps to be executed
	pub number_of_chunks: u32,
	/// The interval in blocks between each swap.
	pub chunk_interval: u32,
}
```

#### Supported Chain and Asset Notation

> Chainflip uses it's own notation for chain and token within our Vault smart contracts. These are the values for the source and destination chains and assets parameters described above. The same values apply for the the corresponding testnets (e.g. Ethereum mainnet and Sepolia share the same value).

Reviewer: ```suggestion
Chainflip uses its own notation for chain and token within our Vault smart contracts. These are the values for the source and destination chains and assets parameters described above. The same values apply for the the corresponding testnets (e.g. Ethereum mainnet and Sepolia share the same value).
```

> | Asset   | Chain      |  Asset ID  |   | Chain      | Chain ID |

Reviewer: This table looks quite ugly

Author: Perhaps we should have thiese as 2 separate tables after all?

---

title: Vault Swaps

description: How to use vault swaps for integration purposes.

---

import { Callout } from "@/components";

## Overview

Traditionally, swaps initiated by Brokers involve requesting a deposit address and sending funds to that deposit address to initiate the swap. This page introduces a new way to initiate swaps on Chainflip by making a direct call to the Chainflip Vault.

> Vault swaps support the same features as those initiated via [Deposit Channels](/concepts/swaps-amm/deposit-channels-and-brokers.mdx#deposit-channels) but don't require a Broker to be initiated.

Reviewer 2: thy do require a broker account
```suggestion
Vault swaps support the same features as those initiated via [Deposit Channels](/concepts/swaps-amm/deposit-channels-and-brokers.mdx#deposit-channels) but don't require to interact with the Chainflip State Chain to be initiated.
```

Reviewer 3: > don't require to interact with the Chainflip State Chain to be initiated

Sadly also not strictly true.

Reviewer 3: LMK what you think: [07a0933](https://github.com/chainflip-io/chainflip-docs/pull/471/commits/07a0933a98d7cdbc1db0c14ca20c12b5270df27c)

Reviewer 2: sound great, thanks!

5. The swaps are scheduled and processed within the Chainflip Protocol's internal AMM system. 

6. The swap output is egressed either into the destination address (simple swaps) or the Smart Contract call for CCM swaps.

### Vault swaps

![Workflow for Swapping via Vault swaps](./vault_swaps.jpeg)

1. Construct and submit a transaction on an external chain. This will be a smart contract call for EVM chains, a program call for Solana, or a particularly encoded transaction for Bitcoin.

* This can be done by manually encoding the transaction (e.g. making the smart contract call) or by using the `request_swap_parameter_encoding` RPC call from our Broker API. It's encouraged to use the Broker API to ensure that the payloads are correctly encoded according to the specifications. Wrongly encoded payloads may result in loss of funds.

2. If using the Broker API, an encoded payload is obtained and must be signed and sent to the appropriate blockchain.

3. The transaction will transfer the appropriate funds to the Chainflip protocol as part of the smart contract call. Chainflip will then process the swap.

The remaining steps are identical to swapping via Deposit Channels.

> ### Pros and Cons

Reviewer 4: I think it's a good idea to have this, especially for integrators. I will add more context.
I'm thinking that maybe we should put this some layer above and highlight a bit more that we have these two ways of swapping with trade-offs. Otherwise this is only displayed when the reader is already into Vault swaps. Wdyt? @fernandezdavid

Reviewer 5: +1 to that @albert-llimos

Reviewer 3: I agree we should highlight pros and cons, but IMO the table is not really appropriate, would be more understandable if we just explain in a few bullet points. For example "Bitcoin CCM support" or "Repeated swaps for the same user" are both quite niche and not very clear for someone without enough context, they need a bit more explanation.

{ 
//TODO: Review this by Albert 
}

The two swap methods each have their own pro and cons. You can use the following information to choose the methods that best suits your situation.

| Considerations | Deposit Channels | Vault swaps |
| -------------- | ---------------- | ----------- | 

> | No. of extrinsic calls   | 2                | 1           |

Reviewer 2: what is meant with extrinsic calls?

| Time it takes to start a swap | > 6 sec | < 1 sec |
| Repeated swaps for the same user | Supported | unsupported |
| Bitcoin CCM support | Yes         | No          |

## Constructing a Vault Swap Transaction

Each of our supported chains encodes data differently. To make this feature easy to use, Chainflip provides the interface to build the transaction on the integrator's behalf. This way the transaction only needs to be signed and is then ready to be sent out to the respective Vault contract.

The function that builds the encoded the transaction data is the [request_swap_parameter_encoding](../../running-a-broker/broker-api/#broker_request_swap_parameter_encoding) RPC call in the Broker API.

<Callout type="info">
The function `request_swap_parameter_encoding()` looks very similar to `request_swap_deposit_address()`, but they are very different functions. 

> - `request_swap_deposit_address()` will submit an extrinsic that opens a deposit channels, and will return the deposit address.

Reviewer 2: ```suggestion
- `request_swap_deposit_address()` will submit an extrinsic that opens a deposit channel, and will return the deposit address.
```

- `request_swap_parameter_encoding()` will **not** make any on-chain changes, but will instead build a transaction that will be returned.
</Callout>

> You may see the use of the `cf_get_vault_swap_details()` RPC call in the documentation (e.g. in example JS code). This call is exactly the same (they share the same underlying logic) as `broker_request_swap_parameter_encoding()` but with the `broker_id` defined explicitly. Within this documentation, you will see the two being used interchangeably. In practice, you can use the RPC call that suits your situation better.

Reviewer 2: `cf_get_vault_swap_details` is called `cf_request_swap_parameter_encoding` now

            "event_data_account": "9acHwMGmeoMr5o8Cw1V2U4HjMQwhced3eQP31yYEhYDU",
            "input_amount": 1000000000,
            "refund_parameters": {
                "retry_duration": 10,
                "refund_address": "BdyHK5DckpAFGcbZveGLPjjMEaADGfNeqcRXKoyd33kA",
                "min_price": "0x0"
            }
        },
        {
          "message": "0x0011223344556677",
          "gas_budget": 1000,
          "ccm_additional_data": "0x"
        }
    ]
}' http://localhost:10997
```

### Construct, sign and broadcast a transaction

With the payload encoded, the next step is to construct a transaction from that payload and sent it to the correct chain's Vault contract address. This step differs for each chain/asset. Please refer to the page dedicated for each chain for more details and other information.

<Callout type="info">
The chain you need to interact with depends on the `Input/Source asset`. 
e.g. if you are swapping ETH into BTC, you should check out the page dedicated to Ethereum and Arbitrum.
</Callout>

[For Bitcoin network](./bitcoin.mdx)

[For Ethereum and Arbitrum](./evm.mdx)

[For Solana](./solana.mdx)

### Payload expiry

> Payloads and therefore transactions for Vault swaps are valid for long periods of time and in some cases indefinitely. However, it is strongly recommended to only sign and broadcast freshly encode payloads and to never rely on previous unused payloads.

Reviewer 2: the bitcoin encoding will be invalid after two epochs, right?

Reviewer 4: Yes exactly. For EVM and Solana it will never expire.

The downside of a vault swap is that the transaction is more complex and will incur higher fees compared to making a simple transfer to a deposit channel.

<Callout type="info">
Learn more in [How Swapping Works](/swapping/how-swapping-works.mdx).
</Callout>

## Workflows

### Swap via Deposit Channels

![Workflow for Swapping via Deposit Channel](./deposit_channel.jpeg)

1. Request a deposit channel via the `pallet_cf_swapping::request_swap_deposit_address_with_affiliates()` extrinsic. This can be done directly or via the `request_swap_deposit_address` RPC call from our Broker API.

2. The generated Deposit address is returned as a response from the RPC call.

3. Deposit fund into the given deposit address.

The remaining steps are shared between both swap workflows.

4. Swaps are processed after deposits are witnessed by Chainflip Engine.

5. The swaps are scheduled and processed within the Chainflip Protocol's internal AMM system. 

6. The swap output is egressed either into the destination address (simple swaps) or the Smart Contract call for CCM swaps.

> ### Vault swaps

Reviewer 3: Bit of a nit, but comparing the two sequence diagrams, The first consider the response with the deposit address to be a step 2. 
In this diagram, we just ignore the reponse from Chainflip. 
Also another nit, it's communicating with a Chainflip Rpc Node, not really with the protocol / network. 

As mentioned elsewhere, it would be good to have these diagrams in an editable state.

![Workflow for Swapping via Vault swaps](./vault_swaps.jpeg)

1. Construct and submit a transaction on an external chain. This will be a smart contract call for EVM chains, a program call for Solana, or a particularly encoded transaction for Bitcoin.

> * This can be done by manually encoding the transaction (e.g. making the smart contract call) or by using the `request_swap_parameter_encoding` RPC call from our Broker API. It's encouraged to use the Broker API to ensure that the payloads are correctly encoded according to the specifications. Wrongly encoded payloads may result in loss of funds.

Reviewer 3: I would rather we just tell people to use the provided RPCs and avoid mentioning that they can build it from scratch - sure they can go it alone but it's non-trivial and trying to explain how is IMO beyond the scope of the docs.

Reviewer 3: They don't need to use the broker API either. But they *will* need to interact with the SC in some way, for example to get affiliate mappings or the current bitcoin vault address.

2. If using the Broker API, an encoded payload is obtained and must be signed and sent to the appropriate blockchain.

3. The transaction will transfer the appropriate funds to the Chainflip protocol as part of the smart contract call. Chainflip will then process the swap.

The remaining steps are identical to swapping via Deposit Channels.

### Pros and Cons

> 

Reviewer 4: Why? I do think we should move it somewhere else as per some comment above but I would have it somewhere. Deposit channels vs Vault swaps has been asked many times.

> 

Reviewer 3: Sorry - doesn't show up correctly in the diff: there was a comment in the file saying 'TODO: review by Albert'

{ 
//TODO: Review this by Albert 
}

The two swap methods each have their own pro and cons. You can use the following information to choose the methods that best suits your situation.

| Considerations | Deposit Channels | Vault swaps |

> | -------------- | ---------------- | ----------- | 

Reviewer 3: I think it's actually more like 12-18 seconds because you need to wait for finality.

| No. of extrinsic calls   | 2                | 1           |
| Time it takes to start a swap | > 6 sec | < 1 sec |
| Repeated swaps for the same user | Supported | unsupported |
| Bitcoin CCM support | Yes         | No          |

> ## Constructing a Vault Swap Transaction

Reviewer 3: This makes it sound a bit easier than it really is... usually there is some additional work to be done by the integrator beyond just signing the return value.

Each of our supported chains encodes data differently. To make this feature easy to use, Chainflip provides the interface to build the transaction on the integrator's behalf. This way the transaction only needs to be signed and is then ready to be sent out to the respective Vault contract.

> The function that builds the encoded the transaction data is the [request_swap_parameter_encoding](../../running-a-broker/broker-api/#broker_request_swap_parameter_encoding) RPC call in the Broker API.

Reviewer 3: Similar to above - i don't think we need to mention this so explicitly. The two methods have different names. Just because the parameters are similar doesn't mean that there's any confusion to clear up. I feel like we'd be better off explaining the difference once, and then concentrating on explaining Vault Swaps here.

<Callout type="info">
The function `request_swap_parameter_encoding()` looks very similar to `request_swap_deposit_address()`, but they are very different functions. 
- `request_swap_deposit_address()` will submit an extrinsic that opens a deposit channels, and will return the deposit address.
- `request_swap_parameter_encoding()` will **not** make any on-chain changes, but will instead build a transaction that will be returned.
</Callout>

> You may see the use of the `cf_get_vault_swap_details()` RPC call in the documentation (e.g. in example JS code). This call is exactly the same (they share the same underlying logic) as `broker_request_swap_parameter_encoding()` but with the `broker_id` defined explicitly. Within this documentation, you will see the two being used interchangeably. In practice, you can use the RPC call that suits your situation better.

Reviewer 3: I don't think we need an example here, I think the sequence diagram/overview is enough. We already have specific examples for each chain in the other sections. No need to repeat here.

</Callout>

```sh
curl -H "Content-Type: application/json" -d '{
    "id": 1,
    "jsonrpc": "2.0",
    "method": "broker_request_swap_parameter_encoding",
    "params": [
		{ "chain": "Solana", "asset": "SOL" },
        { "chain": "Ethereum", "asset": "ETH" },
        "0xc64722AD9613851b10E26fF8118A7696A0f956f2",
        0,
        {
            "chain": "Solana",
            "from": "BdyHK5DckpAFGcbZveGLPjjMEaADGfNeqcRXKoyd33kA",
            "event_data_account": "9acHwMGmeoMr5o8Cw1V2U4HjMQwhced3eQP31yYEhYDU",
            "input_amount": 1000000000,
            "refund_parameters": {
                "retry_duration": 10,
                "refund_address": "BdyHK5DckpAFGcbZveGLPjjMEaADGfNeqcRXKoyd33kA",
                "min_price": "0x0"
            }
        },
        {
          "message": "0x0011223344556677",
          "gas_budget": 1000,
          "ccm_additional_data": "0x"
        }
    ]
}' http://localhost:10997
```

> ### Construct, sign and broadcast a transaction

Reviewer 3: This section is important - should be much more prominent IMO.

Reviewer 4: It's a long and detailed section for each of the chains, I'm not sure we need to explain more here. What else do you want to add?

Reviewer 3: What I mean is it's the part that links to each specific chain implementation. This is the part that integrators need to read most carefully.

[For Solana](./solana.mdx)

### Payload expiry

Payloads and therefore transactions for Vault swaps are valid for long periods of time and in some cases indefinitely. However, it is strongly recommended to only sign and broadcast freshly encode payloads and to never rely on previous unused payloads.

---

### References

#### Input parameters

| Parameter | Description | Data Type |
|-----------|-------------|-----------|
| source_asset | Input asset for the swap | `Asset` |
| destination_asset | Output asset for the swap | `Asset` |
| destination_address | Destination address the swap is paid out to | `AddressString` |
| broker_commission | Broker commission in Basis point | `u16` |
| extra_parameters | Chain specific extra parameters. For details see each chain's dedicated page | `VaultSwapExtraParametersRpc` |
| channel_metadata | Optional: Information for CCM | `Option<CcmChannelMetadata>` |
| boost_fee | Optional: Boost fee for the swap ingress  | `Option<u16>` |
| affiliate_fees | Optional: A list of affiliates and their fees | `Option<Affiliates<AccountId32>>` |
| dca_parameters | Optional: Parameters for DCA swaps | `Option<DcaParameters>` |

> 

Reviewer 3: I think it would make sense to move the detailed type reference to its own page instead of including it on the overview page
